superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py

```diff
@@ -87,6 +97,29 @@ def predictLabelsForItemDetails(self, batchSize, ds, item, model, prog):
         catWeights = tf.nn.softmax(predictions)
         return catWeights, predictions
 
+    def findOptimalBatchSize(self, model, ds) -> int:
+        # Find an optimal batch_size
+        maximum_batchSize: int = len(ds)
+        batchSize: int = 2
+        # We are using a value greater than 0.0 for add_seconds so that small imprecise
+        # timings for small batch sizes don't accidentally trip the time check.
+        add_seconds: float = 0.5
+        previous_time: float = 1e100
+        while batchSize <= maximum_batchSize:
+            try:
+                small_ds = ds.take(batchSize).batch(batchSize)
+                start_time = time.time()
+                model.predict(small_ds, batch_size=batchSize)
+                elapsed_time = time.time() - start_time
+                if elapsed_time > 2 * previous_time + add_seconds:
+                    return batchSize // 2
+                previous_time = elapsed_time
+            except tf.errors.OpError:
+                return batchSize // 2
+            batchSize *= 2
+        # Undo the last doubling; it was spurious
+        return batchSize // 2
```

Comment:
For the case where len(ds) is not a power of two, this will end up returning less than len(ds).  You had been doing the doubling at the start of the while block which would avoid that issue.

Reply:
Thank you, fixing that.